### PR TITLE
New merge features action in feature list's multi selection mode 

### PR DESCRIPTION
**Pull request by nirvn**
This PR adds a new action when multiple features are selected: merge features.

Obligatory GIF:
Peek 2020-07-19 16-43

The merge logic consists of:
- creating a combined geometry from geometry of all selected features
- the _first_ selected feature has its geometry updated to the newly created combined geometry
- all but the first selected features are subsequently removed (i.e., the 'merged' feature retains the attributes of the first selected feature)

### CODE REVIEW

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-2666e334cccef12930aaca40f1346db7b4d1e0bc-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-2666e334cccef12930aaca40f1346db7b4d1e0bc-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-2666e334cccef12930aaca40f1346db7b4d1e0bc-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-2666e334cccef12930aaca40f1346db7b4d1e0bc-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-252a3b7eb6962cb15ca9b2bfe5adc40a18da020e-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-252a3b7eb6962cb15ca9b2bfe5adc40a18da020e-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-252a3b7eb6962cb15ca9b2bfe5adc40a18da020e-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-252a3b7eb6962cb15ca9b2bfe5adc40a18da020e-x86_64.apk)

**Comment by nirvn** [comment]
Instead of resetting the whole model on feature deletion, we now rely on removeRows and its pair of begin/end signal. 

Dividends of doing-the-right-thing(tm): while testing this, I've realized we were until now failing to emit a dataChanged signal whenever an attributeValueChanged signal was triggered. This led, among other things, to the feature display name being out of sync when editing the field (or any field part of the display expression). I had seen that bug before and couldn't diagnose it then.

:tada:

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-205d3603b45efcda80962c444fbfc11056dbe998-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-205d3603b45efcda80962c444fbfc11056dbe998-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-205d3603b45efcda80962c444fbfc11056dbe998-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-205d3603b45efcda80962c444fbfc11056dbe998-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-01fee215d9aa0251c12cfb7a58959b782e43c3dc-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-01fee215d9aa0251c12cfb7a58959b782e43c3dc-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-01fee215d9aa0251c12cfb7a58959b782e43c3dc-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-01fee215d9aa0251c12cfb7a58959b782e43c3dc-x86_64.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-659b8d0e2cef2928638b1714a886281060de4e8c-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-659b8d0e2cef2928638b1714a886281060de4e8c-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-659b8d0e2cef2928638b1714a886281060de4e8c-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-659b8d0e2cef2928638b1714a886281060de4e8c-x86_64.apk)

**Comment by nirvn** [comment]
@m-kuhn , this is ready to review / approve.

File: src/core/multifeaturelistmodel.h [hunk 1]
```diff
@@ -72,6 +73,13 @@ class MultiFeatureListModel : public QSortFilterProxyModel
      */
     Q_INVOKABLE void clearSelection();
 
+    /**
+     * Returns the display name of a given feature.
+     * \param vlayer the vector layer containing the feature
+     * \param feature the display name feature
+     */
+    Q_INVOKABLE QString featureDisplayName( QgsVectorLayer *vlayer, QgsFeature feature );
```
Comment (m-kuhn) [inline src/core/multifeaturelistmodel.h:81]:
```suggestion
    Q_INVOKABLE QString featureDisplayName( QgsVectorLayer *vlayer, const QgsFeature &feature ) const;
```

... and evenmore: does this belong in here -- or rather into a singleton/static "utils" class?

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-31fb77f2dd43dd8829b8bb897128f86490031fd6-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-31fb77f2dd43dd8829b8bb897128f86490031fd6-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-31fb77f2dd43dd8829b8bb897128f86490031fd6-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-31fb77f2dd43dd8829b8bb897128f86490031fd6-x86_64.apk)

File: src/core/multifeaturelistmodel.h [hunk 2]
```diff
@@ -82,14 +90,24 @@ class MultiFeatureListModel : public QSortFilterProxyModel
...
 
+    //! Returns TRUE if the selected features can be merged
+    bool canMergeSelection();
+
     //! Returns TRUE if the selected features can be deleted
     bool canDeleteSelection();
 
+    /**
+     * Merges selected features by updating the first seleted feature's geometry
+     * to a combinasion (i.e. union) of geometries of all selected features.
```
Comment (m-kuhn) [inline src/core/multifeaturelistmodel.h:104]:
```suggestion
     * to a combination (i.e. union) of geometries of all selected features.
```

combinaison?

File: src/core/multifeaturelistmodelbase.cpp [hunk 3]
```diff
@@ -316,22 +340,95 @@ bool MultiFeatureListModelBase::canDeleteSelection()
...
+    if ( combinedGeometry.lastError().isEmpty() )
+    {
+      combinedGeometry = geom;
+    }
+    else
+    {
+      isSuccess = false;
+      break;
+    }
+    qDebug() << combinedGeometry.asWkt();
```
Comment (m-kuhn) [inline src/core/multifeaturelistmodelbase.cpp:370]: Remove before merge?

File: src/core/multifeaturelistmodelbase.cpp [hunk 4]
```diff
@@ -316,22 +340,95 @@ bool MultiFeatureListModelBase::canDeleteSelection()
...
+    }
+
+    QgsFeature mergedFeature = selectedFeatures[0].second;
+    mergedFeature.setGeometry( combinedGeometry );
+    isSuccess = vlayer->updateFeature( mergedFeature );
+
+    if ( isSuccess )
+    {
+      selectedFeatures.removeFirst();
+      for ( const auto &pair : selectedFeatures )
```
Comment (m-kuhn) [inline src/core/multifeaturelistmodelbase.cpp:388]:
```suggestion
      for ( const auto &pair : qgis::as_const( selectedFeatures ) )
```

File: src/core/multifeaturelistmodelbase.cpp [hunk 5]
```diff
@@ -543,3 +640,32 @@ void MultiFeatureListModelBase::attributeValueChanged( QgsFeatureId fid, int idx
...
   }
 }
+
+void MultiFeatureListModelBase::geometryChanged( QgsFeatureId fid, const QgsGeometry &geometry )
+{
+  QgsVectorLayer *l = qobject_cast<QgsVectorLayer *>( sender() );
+  Q_ASSERT( l );
+
+  int i = 0;
+  for ( auto &pair : mFeatures )
```
Comment (m-kuhn) [inline src/core/multifeaturelistmodelbase.cpp:650]:
```suggestion
  for ( auto &pair : qgis::as_const( mFeatures ) )
```

or can the method be made const?

**Comment by m-kuhn** [comment]
Good work !

File: src/core/multifeaturelistmodel.h (hunk 2 above)
Comment (nirvn) [reply src/core/multifeaturelistmodel.h:104]: My head was evidentially having an internal fight between French and English, must have been a draw :wink:

File: src/core/multifeaturelistmodelbase.cpp (hunk 5 above)
Comment (nirvn) [reply src/core/multifeaturelistmodelbase.cpp:650]: Can it be const if we are modifying list items?

Comment (nirvn) [reply src/core/multifeaturelistmodelbase.cpp:650]: Can this (or the method) be const if we're modifying lists elements?

**Comment by qfield-fairy** [comment]
Uploaded test apks for
 - [**armv7**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-274b0a487a98833efd8b69c1a8f37cd89c9f4fa0-armv7.apk)
 - [**x86**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-274b0a487a98833efd8b69c1a8f37cd89c9f4fa0-x86.apk)
 - [**arm64_v8a**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-274b0a487a98833efd8b69c1a8f37cd89c9f4fa0-arm64_v8a.apk)
 - [**x86_64**](https://download.opengis.ch/qfield/ci-builds/qfield-dev-1144-274b0a487a98833efd8b69c1a8f37cd89c9f4fa0-x86_64.apk)

**Comment by nirvn** [comment]
@m-kuhn , comments addressed.

**Comment by nirvn** [comment]
:tada:

**Comment by signedav** [comment]
This is only possible on multipolygon layers, right @nirvn?

**Comment by nirvn** [comment]
@signedav , that's correct.

**Pull request merged**